scipy/optimize/_numdiff.py

```diff
@@ -344,34 +363,77 @@ def fun_wrapped(x):
     if np.any((x0 < lb) | (x0 > ub)):
         raise ValueError("`x0` violates bound constraints.")
 
-    h = _compute_absolute_step(rel_step, x0, method)
+    if return_linear_operator:
+        if rel_step is None:
+            rel_step = _compute_relative_step(method)
+
+        return _linear_operator_difference(fun_wrapped, x0,
+                                           f0, rel_step, method)
+    else:
+        h = _compute_absolute_step(rel_step, x0, method)
+
+        if method == '2-point':
+            h, use_one_sided = _adjust_scheme_to_bounds(
+                x0, h, 1, '1-sided', lb, ub)
+        elif method == '3-point':
+            h, use_one_sided = _adjust_scheme_to_bounds(
+                x0, h, 1, '2-sided', lb, ub)
+        elif method == 'cs':
+            use_one_sided = False
+
+        if sparsity is None:
+            return _dense_difference(fun_wrapped, x0, f0, h,
+                                     use_one_sided, method)
+        else:
+            if not issparse(sparsity) and len(sparsity) == 2:
+                structure, groups = sparsity
+            else:
+                structure = sparsity
+                groups = group_columns(sparsity)
+
+            if issparse(structure):
+                structure = csc_matrix(structure)
+            else:
+                structure = np.atleast_2d(structure)
+
+            groups = np.atleast_1d(groups)
+            return _sparse_difference(fun_wrapped, x0, f0, h,
+                                      use_one_sided, structure,
+                                      groups, method)
+
+
+def _linear_operator_difference(fun, x0, f0, dx, method):
+    m = f0.size
+    n = x0.size
 
     if method == '2-point':
-        h, use_one_sided = _adjust_scheme_to_bounds(
-            x0, h, 1, '1-sided', lb, ub)
+        def matvec(p):
+            if np.array_equal(p, np.zeros_like(p)):
+                return np.zeros(m)
+            x = x0 + dx * p
+            print(x0, p, dx, x - x0)
+            df = fun(x) - f0
+            return df/dx
```

Comment:
I think the conventional approach is to recomputed `dx` from `x` and `x0` to make `df` and `dx` match precisely.